tuxemon/core/components/networking.py

```diff
@@ -0,0 +1,777 @@
+                self.keys[pg.K_LEFT] = 1
+                event = self.game.keyboard_events["KEYDOWN"]["left"]
+
+            elif event_data == "KEYUP:left":
+                self.keys[pg.K_LEFT] = 0
+                event = self.game.keyboard_events["KEYUP"]["left"]
+
+            elif event_data == "KEYDOWN:right":
+                self.keys[pg.K_RIGHT] = 1
+                event = self.game.keyboard_events["KEYDOWN"]["right"]
+
+            elif event_data == "KEYUP:right":
+                self.keys[pg.K_RIGHT] = 0
+                event = self.game.keyboard_events["KEYUP"]["right"]
+
+            elif event_data == "KEYDOWN:enter":
+                self.keys[pg.K_RETURN] = 1
+                event = self.game.keyboard_events["KEYDOWN"]["enter"]
+
+            elif event_data == "KEYUP:enter":
+                self.keys[pg.K_RETURN] = 0
+                event = self.game.keyboard_events["KEYUP"]["enter"]
+
+            elif event_data == "KEYDOWN:esc":
+                self.keys[pg.K_ESCAPE] = 1
+                event = self.game.keyboard_events["KEYDOWN"]["escape"]
+
+            elif event_data == "KEYUP:esc":
+                self.keys[pg.K_ESCAPE] = 0
+                event = self.game.keyboard_events["KEYUP"]["escape"]
+            
+            else:
+                logger.debug("Unknown network event: " +str(event_data))
+                event = None
+
+            if event:
+                events.append(event)
+        
+        # Clear out the network events list once all events have been processed.
+        self.network_events = []
+        return events
+    
+    
+    def server_event_handler(self, cuuid, event_data):
+        """Handles events sent from the middleware that are legal.
+
+        :param cuuid: Clients unique user identification number.
+        :param event_data: Event information sent by client.
+        
+        :type cuuid: String 
+        :type event_data: Dictionary
+
+        :rtype: None
+        :returns: None
+
+        """
+        # Only respond to the latest message of a given type
+        if not "event_list" in self.server.registry[cuuid]:
+            self.server.registry[cuuid]["event_list"] = {}
+        elif not event_data["type"] in self.server.registry[cuuid]["event_list"]:
+                self.server.registry[cuuid]["event_list"][event_data["type"]] = -1
+        
+        elif event_data["event_number"] <= self.server.registry[cuuid]["event_list"][event_data["type"]]:
+            return False
+        else:
+            self.server.registry[cuuid]["event_list"][event_data["type"]] = event_data["event_number"]
+                
+        if event_data["type"] == "PUSH_SELF":
+            self.server.registry[cuuid]["sprite_name"] = event_data["sprite_name"]
+            self.server.registry[cuuid]["map_name"] = event_data["map_name"]
+            self.server.registry[cuuid]["char_dict"] = event_data["char_dict"]
+            self.notify_populate_client(cuuid, event_data)
```

Comment:
Agreed. This method of multiplayer would enable people to do multiplayer interaction and play with their local save game. Pushing this to a server will force you to make a new network only character. Once a network infrastructure is in place that can be targeted for development this will be implemented.